### ISSUE TITLE
ENH: Use Ryu

### ISSUE THREAD

**Issue by HaoZeke** [HaoZeke/fastMatMR#20]
Or another floating point library. As is done upstream.

**Comment by alugowski** [comment]
Pardon if this is a silly question, but I'm not up to date on how R extensions are built. How much control do you have over what compiler is used? I ask because all major compilers' latest versions basically bundle Ryu as part of the standard library. The reason FMM vendors Ryu is to support older compilers in wide use that do not. Basically if you can use `std::to_chars` with floats then you're probably already using Ryu.

fast_float for the read side might have more value. It's also getting bundled with compilers, but not all yet. MSVC is still in PR I believe, and LLVM only uses it for the C methods and not for `std::from_chars`.  On the other hand fast_float is a single header, so integrating it is very easy compared to Ryu. Just add the header to the include dir and define `FMM_USE_FAST_FLOAT`.

**Comment by HaoZeke** [comment]
> Pardon if this is a silly question, but I'm not up to date on how R extensions are built. How much control do you have over what compiler is used? I ask because all major compilers' latest versions basically bundle Ryu as part of the standard library. The reason FMM vendors Ryu is to support older compilers in wide use that do not. Basically if you can use `std::to_chars` with floats then you're probably already using Ryu.

That's a great question. For packages which are meant for CRAN (the Comprehensive R Archive Network, like PyPi for R), it is generally safe to assume GCC 10 (as part of the [Rtools42](https://cran.r-project.org/bin/windows/Rtools/rtools42/rtools.html) compiler toolchain), though up-to-date / current versions should have GCC 12 from [Rtools43](https://cran.r-project.org/bin/windows/Rtools/rtools43/rtools.html). I think GCC 10 only had integer support for `std::to_chars` but I might be wrong.

> fast_float for the read side might have more value. It's also getting bundled with compilers, but not all yet. MSVC is still in PR I believe, and LLVM only uses it for the C methods and not for `std::from_chars`. On the other hand fast_float is a single header, so integrating it is very easy compared to Ryu. Just add the header to the include dir and define `FMM_USE_FAST_FLOAT`.

Makes sense, will do, thank you :)

**Comment by alugowski** [comment]
Yeah that's old enough to need Ryu/Dragonbox/fast_float. The upside is that you'll see some very large speedup, especially on 4+ cores.

Ryu isn't too hard to add either. Just a few .c files to add to your target, or build and link a small library.

**Comment by alugowski** [comment]
> GCC 12 from [Rtools43](https://cran.r-project.org/bin/windows/Rtools/rtools43/rtools.html).

I checked again, and GCC 12 is the first one with the fast methods bundled. So if you're able to use GCC 12 then there's nothing more to do :)
